balancer.go

```diff
@@ -0,0 +1,337 @@
+	// to make load balancing decision. This is from the metadata attached
+	// in the address updates from name resolver.
+	Metadata interface{}
+}
+
+// BalancerGetOptions configures a Get call.
+// This is the EXPERIMENTAL API and may be changed or extended in the future.
+type BalancerGetOptions struct {
+	// BlockingWait specifies whether Get should block when there is no
+	// connected address.
+	BlockingWait bool
+}
+
+// Balancer chooses network addresses for RPCs.
+// This is the EXPERIMENTAL API and may be changed or extended in the future.
+type Balancer interface {
+	// Start does the initialization work to bootstrap a Balancer. For example,
+	// this function may start the name resolution and watch the updates.
+	Start(target string) error
+	// Up informs the Balancer that gRPC has a connection to the server at
+	// addr. It returns down which is called once the connection to addr gets
+	// lost or closed.
+	Up(addr Address) (down func(error))
+	// Get gets the address of a server for the RPC corresponding to ctx.
+	// i) If it returns a connected address, gRPC internals issues the RPC on the
+	// connection to this address;
+	// ii) If it returns an address on which the connection is under construction
+	// (initiated by Notify(...)) but not connected, gRPC internals
+	//  * fails RPC if the RPC is fail-fast and connection is in the TransientFailure
+	//  or Shutdown state;
+	//  or
+	//  * issues RPC on the connection otherwise.
+	// iii) If it returns an address on which the connection does not exist, gRPC
+	// internals treats it as an error and will fail the corresponding RPC.
+	//
+	// Therefore, we recommend the following general rule when you write your own
+	// custom Balancer.
+	//
+	// If opts.BlockingWait is true, it should return a connected address or
+	// block if there is no connected address. It should respect the timeout or
+	// cancellation of ctx when blocking. If opts.BlockingWait is false (for fail-fast
+	// RPCs), it should return an address it has notified via Notify(...) immediately
+	// instead of blocking.
+	//
+	// The function returns put which is called once the rpc has completed or failed.
+	// put can collect and report RPC stats to a remote load balancer.
+	Get(ctx context.Context, opts BalancerGetOptions) (addr Address, put func(), err error)
+	// Notify notifies gRPC internals the list of Address to be connected. The list
+	// may be from a name resolver or remote load balancer. gRPC internals will
+	// compare it with the existing connected addresses. If the address Balancer
+	// notified is not in the existing connected addresses, gRPC starts to connect
+	// the address. If an address in the existing connected addresses is not in
+	// the notification list, the corresponding connection is shutdown gracefully.
+	// Otherwise, there are no operations to take. Note that this function must
+	// return the full list of the Addresses which should be connected. It is NOT delta.
+	Notify() <-chan []Address
```

Comment:
On Fri, May 27, 2016 at 7:24 AM, Zellyn Hunter <notifications@github.com>
wrote:

> In balancer.go
> <https://github.com/grpc/grpc-go/pull/690#discussion_r64911998>:
>
> > +	// cancellation of ctx when blocking. If opts.BlockingWait is false (for fail-fast
> > +	// RPCs), it should return an address it has notified via Notify(...) immediately
> > +	// instead of blocking.
> > +	//
> > +	// The function returns put which is called once the rpc has completed or failed.
> > +	// put can collect and report RPC stats to a remote load balancer.
> > +	Get(ctx context.Context, opts BalancerGetOptions) (addr Address, put func(), err error)
> > +	// Notify notifies gRPC internals the list of Address to be connected. The list
> > +	// may be from a name resolver or remote load balancer. gRPC internals will
> > +	// compare it with the existing connected addresses. If the address Balancer
> > +	// notified is not in the existing connected addresses, gRPC starts to connect
> > +	// the address. If an address in the existing connected addresses is not in
> > +	// the notification list, the corresponding connection is shutdown gracefully.
> > +	// Otherwise, there are no operations to take. Note that this function must
> > +	// return the full list of the Addresses which should be connected. It is NOT delta.
> > +	Notify() <-chan []Address
>
> Called repeatedly? So, on the first call, you return a chan that you'll
> write to whenever the Balancer wants to notify grpc internals of changes.
> On the second call, you return a different chan that…? Or are you expected
> to write to the first chan once, then close it?
>
By "called repeatedly", I meant something like

for {
  addrs, ok := <-lb.Notify()
  if !ok {
    break
  }
}

Whether it returns the same chan or not is implementation detail. I believe
most of impls should return the same chan always (e.g., roundRobin Balancer
I made).

>
>
—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc-go/pull/690/files/db887c9bcd224a919fae77c63283655504aa2fcd#r64911998>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJpyLDwyvyW5W5EcwrVr1YIgPvpl-H9Cks5qFv6DgaJpZM4Ih1rh>
> .
>